### PR TITLE
document type bound operators

### PR DESCRIPTION
**Pull request by timotheecour**
refs https://github.com/nim-lang/Nim/pull/17062#issuecomment-780192544

just reflects my understanding of it, please correct if needed

## future work
- [x] document `=trace` and `=dispose`, even if it comes with a disclaimer that these may change

### CODE REVIEW

File: doc/manual.rst [hunk 1]
```diff
@@ -3678,6 +3678,47 @@ Is short for:
 
 
 
+Routines
+--------
+
+A routine is a symbol of kind: `proc`, `func`, `iterator`, `macro`, `template`, `converter`.
```
Comment (Araq) [inline doc/manual.rst]: While this is indeed the definition of a "routine" (except that you forgot to mention ``method``) the type bound operations cannot be any "routine", unsupported are: iterators, macros, methods, templates and converters.

File: doc/manual.rst [hunk 2]
```diff
@@ -3678,6 +3678,47 @@ Is short for:
...
+  block:
+    var a = initFoo(int)
+    doAssert witness == 1
+    `=destroy`(a) # can be called explicitly, even without being in scope
+    doAssert witness == 2
+  # will still be called upon exiting scope
+  doAssert witness == 3
+
+Type bound routines allowed names are in ``ast.AttachedOpToStr``, currently include:
+``=destroy``, ``=copy``, ``=sink``, ``=trace``, ``=dispose``, ``=deepcopy``.
```
Comment (Araq) [inline doc/manual.rst]: Btw ``=trace`` and ``=dispose`` need to be documented but ideally only a potentially destructive ``=trace`` is needed and no ``=dispose``. It's currently entirely unclear what performance impact this has for ``--gc:orc`` which is why this was not documented.

File: doc/manual.rst (hunk 1 above)
Comment (timotheecour) [reply doc/manual.rst]: fixed; indeed, I verified that `template` didn't work

File: doc/manual.rst (hunk 2 above)
Comment (timotheecour) [reply doc/manual.rst]: deferred to future work, but I've changed wording to say that some of those are implementation defined

**Comment by timotheecour** [comment]
@Araq  PTAL

File: doc/manual.rst [hunk 3]
```diff
@@ -3678,6 +3678,48 @@ Is short for:
...
+  # main.nim:
+  import foo
+  block:
+    var a = initFoo(int)
+    doAssert witness == 0
+  doAssert witness == 1
+  block:
+    var a = initFoo(int)
+    doAssert witness == 1
+    `=destroy`(a) # can be called explicitly, even without being in scope
```
Comment (konsumlamm) [inline doc/manual.rst:3711]: This comment sounds weird to me, how is anything "not in scope"? `=destroy` and `a` clearly are in scope here. Did you mean "without being at the end of a scope" perhaps?

Comment (timotheecour) [reply doc/manual.rst:3711]: what i meant is `foo.=destroy` is not in scope according to regular symbol resolution rules, it's a private symbol in foo.nim and here we're in main.nim; if it were named, say `destroy`, it wouldn't work.

I could rephrase this as: "can be called explicitly, as it's implicitly in scope despite `foo.=destroy` being private"

Comment (konsumlamm) [reply doc/manual.rst:3711]: Oh right. Ye, it would be good to clarify this.

File: doc/manual.rst [hunk 4]
```diff
@@ -3678,6 +3678,48 @@ Is short for:
...
+    doAssert witness == 2
+  # will still be called upon exiting scope
+  doAssert witness == 3
+
+Type bound operators allowed names are in ``ast.AttachedOpToStr``, currently include:
```
Comment (Araq) [inline doc/manual.rst]:
```suggestion
Type bound operators currently include:
```
